src/Data/List/Types.purs

```diff
@@ -65,7 +64,47 @@ instance monoidList :: Monoid (List a) where
   mempty = Nil
 
 instance functorList :: Functor List where
-  map f = foldr (\x acc -> f x : acc) Nil
+  map = listMap
+
+-- chunked list Functor inspired by OCaml
+-- https://discuss.ocaml.org/t/a-new-list-map-that-is-both-stack-safe-and-fast/865
+-- chunk sizes determined through experimentation
+listMap :: forall a b. (a -> b) -> List a -> List b
+listMap f = startUnrolledMap unrollLimit
+  where
+  -- iterate the unrolled map up to 1000 times,
+  -- which hits up to 5000 elements
+  unrollLimit = 1000
+
+  startUnrolledMap :: Int -> List a -> List b
+  startUnrolledMap 0 (x : xs) = f x : chunkedRevMap xs
+  startUnrolledMap n (x1 : x2 : x3 : x4 : x5 : xs) =
+    f x1 : f x2 : f x3 : f x4 : f x5 : startUnrolledMap (n - 1) xs
```

Comment:
I'm not sure if OCaml has tail calls modulo cons, but isn't this going to lead to a stack overflow for us?

Reply:
This is why I've limited the "unrolled map" to 1000 iterations. It should avoid stack overflow. Are you suggesting that there is a way to make it stack overflow despite that limit?

Reply:
I guess not, unless `f` was already on the verge of overflowing the stack. It can only make an existing problem a finite amount worse :)